!htc HTTPS-enable server- and client-side APIs

sirthias opened the pull request:
This extends the akka-http-core APIs with optional support for HTTPS connection.
Currently there is no deeper support for actively reacting to or triggering session renegotiations but everything should be working nicely.

I have updated the `akka.http.scaladsl.TestClient` in _akka-http-core_ to talk to `https://github.com`, which works on the connection-level but somehow fails when going through a pool.
I'll try to dig up the problem and add a fix to this PR, but this shouldn't keep you from reviewing. 

/cc @jrudolph

rkuhn commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala:
```diff
@@ -87,10 +88,12 @@ private[http] object OutgoingConnectionBlueprint {
-      val logger = Flow[ByteString].transform(() ⇒ errorLogger(log, "Outgoing request stream error")).named("errorLogger")
-      val bytesOut = (terminationMerge.out ~> requestRendering.via(logger)).outlet
+      val logger = b.add(Flow[ByteString].transform(() ⇒ errorLogger(log, "Outgoing request stream error")).named("errorLogger"))
+      val wrapTls = b.add(Flow[ByteString].map[SslTlsOutbound](SendBytes))
```
Does this not work without the type ascription? That would point to a missing variance annotation somewhere …

sirthias replied:
It also work without the ascription.
Appears to be a left-over from one of my fights with the DSL.
I'll remove it.

rkuhn commented on akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala:
```diff
@@ -465,3 +570,39 @@ object Http extends ExtensionId[HttpExt] with ExtensionIdProvider {
+}
+
+object HttpsContext {
+  /** Java API */
```
The Java API is in the javadsl package already, isn’t it?

sirthias replied:
Yes, this should be `private[http]` and marked as internal.

rkuhn commented on akka-stream/src/main/scala/akka/stream/io/SslTls.scala:
```diff
@@ -49,6 +49,9 @@ import java.security.cert.Certificate
 object SslTls {
 
+  type ScalaFlow = scaladsl.BidiFlow[SslTlsOutbound, ByteString, ByteString, SslTlsInbound, Unit]
+  type JavaFlow = javadsl.BidiFlow[SslTlsOutbound, ByteString, ByteString, SslTlsInbound, Unit]
```
yes, indeed

rkuhn commented:
This looks quite good as a first step, thanks!

jrudolph commented:
LGTM

jrudolph commented on akka-stream/src/main/scala/akka/stream/io/SslTls.scala:
```diff
@@ -138,14 +141,14 @@ object SslTls {
  */
 object SslTlsPlacebo {
-  val forScala = scaladsl.BidiFlow() { implicit b ⇒
+  val forScala: SslTls.ScalaFlow = scaladsl.BidiFlow() { implicit b ⇒
```
This made the type more generic than it used to be leading to the test compilation failure: 

before its type was

```scala
scaladsl.BidiFlow[SslTlsOutbound, ByteString, ByteString, SessionBytes, Unit]
```

now it's

```scala
scaladsl.BidiFlow[SslTlsOutbound, ByteString, ByteString, SslTlsInbound, Unit]
```

sirthias replied:
Yes, the type is now wider, which I'd consider to be better.
Which test compilation failure do you mean?

jrudolph replied:
In TlsSpec:

```
[info] Compiling 88 Scala sources and 13 Java sources to /mnt/akka/jenkinsakka/localhome-a4/workspace/pr-validator-per-commit-jenkins/akka-stream-tests/target/test-classes...
[error] /mnt/akka/jenkinsakka/localhome-a4/workspace/pr-validator-per-commit-jenkins/akka-stream-tests/src/test/scala/akka/stream/io/TlsSpec.scala:442: value bytes is not a member of akka.stream.io.SslTlsInbound
[error]       result.map(_.bytes) should be((1 to 3).map(b ⇒ ByteString(b.toByte)))
[error]                    ^
[error] one error found
```

sirthias replied:
Ah, ok.
I'll narrow it down again then.
Thanks!

sirthias committed:
!htc HTTPS-enable server- and client-side APIs

sirthias commented:
Ok, incorporated the review feedback and fixed the small bug preventing pooled HTTPS access to github.com yesterday.
Everything runs smoothly now on the HTTP layer.

However, there apparently there is one problem left in the TLS or the IO layer:

```
akka-http-core-experimental > test:runMain akka.http.scaladsl.TestClient
[info] Running akka.http.scaladsl.TestClient
[DEBUG] [05/13/2015 14:46:39.160] [run-main-27] [EventStream(akka://ServerTest)] logger log1-Logging$DefaultLogger started
[DEBUG] [05/13/2015 14:46:39.162] [run-main-27] [EventStream(akka://ServerTest)] Default Loggers started
Fetching HTTPS server version of host `github.com` via a direct low-level connection ...
[DEBUG] [05/13/2015 14:46:39.775] [ServerTest-akka.actor.default-dispatcher-4] [akka://ServerTest/system/IO-TCP/selectors/$a/0] Attempting connection to [github.com/192.30.252.128:443]
[DEBUG] [05/13/2015 14:46:39.901] [ServerTest-akka.actor.default-dispatcher-4] [akka://ServerTest/system/IO-TCP/selectors/$a/0] Connection established to [github.com/192.30.252.128:443]
[WARN] [05/13/2015 14:46:40.330] [ServerTest-akka.test.stream-dispatcher-9] [ActorSystem(ServerTest)] Illegal response header: Illegal 'set-cookie' header: Invalid input '.', expected '\r', WSP or ALPHANUM (line 1, column 22): logged_in=no; domain=.github.com; path=/; expires=Sun, 13-May-2035 12:46:39 GMT; secure; HttpOnly
                     ^
github.com is running Server: GitHub.com

Fetching HTTP server version of host `github.com` via the high-level API ...
[DEBUG] [05/13/2015 14:46:40.380] [ServerTest-akka.actor.default-dispatcher-2] [akka://ServerTest/user/PoolInterfaceActor-0] (Re-)starting host connection pool to github.com:443
[DEBUG] [05/13/2015 14:46:40.451] [ServerTest-akka.actor.default-dispatcher-3] [akka://ServerTest/system/IO-TCP/selectors/$a/1] Attempting connection to [github.com/192.30.252.128:443]
[WARN] [05/13/2015 14:46:40.548] [ServerTest-akka.actor.default-dispatcher-3] [akka://ServerTest/user/event-stream-logger] UnhandledMessage(ResumeReading,Actor[akka://ServerTest/system/IO-TCP-STREAM/client-1-github.com%2F192.30.252.128%3A443#1626416480],Actor[akka://ServerTest/system/IO-TCP/selectors/$a/0#-76700019])
[DEBUG] [05/13/2015 14:46:40.569] [ServerTest-akka.actor.default-dispatcher-3] [akka://ServerTest/system/IO-TCP/selectors/$a/1] Connection established to [github.com/192.30.252.128:443]
[WARN] [05/13/2015 14:46:40.954] [ServerTest-akka.test.stream-dispatcher-12] [ActorSystem(ServerTest)] Illegal response header: Illegal 'set-cookie' header: Invalid input '.', expected '\r', WSP or ALPHANUM (line 1, column 22): logged_in=no; domain=.github.com; path=/; expires=Sun, 13-May-2035 12:46:40 GMT; secure; HttpOnly
                     ^
github.com is running Server: GitHub.com
[INFO] [05/13/2015 14:46:40.963] [ServerTest-akka.actor.default-dispatcher-3] [ActorSystem(ServerTest)] Initiating orderly shutdown of all active host connections pools...
[DEBUG] [05/13/2015 14:46:40.965] [ServerTest-akka.actor.default-dispatcher-3] [akka://ServerTest/user/PoolInterfaceActor-0] Shutting down host connection pool to github.com:443
[DEBUG] [05/13/2015 14:46:45.368] [ServerTest-akka.test.stream-dispatcher-8] [akka://ServerTest/user/$a/flow-3-4-publisherSource-prefixAndTail] Cancelling akka.stream.impl.MultiStreamOutputProcessor$SubstreamOutput@3219165a (after: 5000 ms)
[DEBUG] [05/13/2015 14:46:45.968] [ServerTest-akka.test.stream-dispatcher-9] [akka://ServerTest/user/$a/flow-7-2-publisherSource-prefixAndTail] Cancelling akka.stream.impl.MultiStreamOutputProcessor$SubstreamOutput@238cd025 (after: 5000 ms)
[DEBUG] [05/13/2015 14:46:45.973] [ServerTest-akka.actor.default-dispatcher-2] [akka://ServerTest/user/SlotProcessor-2] Slot 0 disconnected after regular connection close
[DEBUG] [05/13/2015 14:46:45.974] [ServerTest-akka.actor.default-dispatcher-14] [akka://ServerTest/user/PoolInterfaceActor-0] Host connection pool to github.com:443 has completed orderly shutdown
[INFO] [05/13/2015 14:46:45.976] [ServerTest-akka.actor.default-dispatcher-4] [ActorSystem(ServerTest)] STOPPED
[DEBUG] [05/13/2015 14:46:45.987] [ServerTest-akka.actor.default-dispatcher-2] [EventStream] shutting down: StandardOutLogger started
[success] Total time: 8 s, completed May 13, 2015 2:46:46 PM
```

The run completes successfully but the tear-down of the connection pool takes 4+ seconds (as opposed to a few ms when run w/o encryption). Note the suspicous

```
UnhandledMessage(ResumeReading,Actor[akka://ServerTest/system/IO-TCP-STREAM/client-1-github.com%2F192.30.252.128%3A443#1626416480],Actor[akka://ServerTest/system/IO-TCP/selectors/$a/0#-76700019])
```

In the log. Anybody has an idea of what might be going on here?

drewhk commented:
LGTM, I think we can merge and investigate the IO layer problem in a separate issue

sirthias commented:
Ok, great!

drewhk commented:
Ticket to investigate TLS behavior: https://github.com/akka/akka/issues/17504

wsargent commented:
:+1:
